src/many-modules/src/_9_account/features/multisig.rs

```diff
@@ -9,98 +9,168 @@ use many_macros::many_module;
+                        "Data size ({}) over limit ({})",
+                        value.len(),
+                        M
+                    )));
+                }
+                Ok($item(value.into()))
+            }
+        }
+        )*
+    };
+}
+
+declare_try_from!(
+    String = Self::String;
+    &str = Self::String;
+    ByteVec = Self::ByteString;
+    Vec<u8> = Self::ByteString;
+);
+
+impl<const M: usize> TryFrom<Either<String, ByteVec>> for MemoInner<M> {
+    type Error = ManyError;
+
+    fn try_from(value: Either<String, ByteVec>) -> Result<Self, Self::Error> {
+        match value {
+            Either::Left(str) => Self::try_from(str),
+            Either::Right(bstr) => Self::try_from(bstr),
+        }
     }
 }
 
-impl<C> Decode<'_, C> for Memo<String> {
-    fn decode(d: &mut Decoder<'_>, ctx: &mut C) -> Result<Self, decode::Error> {
-        let Memo(s): Memo<&str> = Memo::decode(d, ctx)?;
-        Ok(Memo(s.into()))
+impl<C, const M: usize> Encode<C> for MemoInner<M> {
+    fn encode<W: encode::Write>(
+        &self,
+        e: &mut Encoder<W>,
+        _: &mut C,
+    ) -> Result<(), encode::Error<W::Error>> {
+        match self {
+            MemoInner::String(str) => e.str(str),
+            MemoInner::ByteString(bstr) => e.bytes(bstr.as_slice()),
+        }
+        .map(|_| ())
     }
 }
 
-// Implementing for completeness, in case we want zero-allocation parsing in the future
-impl<'b, C> Decode<'b, C> for Memo<&'b str> {
+impl<'b, C, const M: usize> Decode<'b, C> for MemoInner<M> {
     fn decode(d: &mut Decoder<'b>, _: &mut C) -> Result<Self, decode::Error> {
-        let s = d.str()?;
-        if s.as_bytes().len() > MULTISIG_MEMO_DATA_MAX_SIZE {
-            return Err(decode::Error::message("Memo size over limit"));
+        match d.datatype()? {
+            Type::Bytes => Self::try_from(d.bytes()?.to_vec()).map_err(decode::Error::message),
+            Type::String => Self::try_from(d.str()?).map_err(decode::Error::message),
+            // Type::BytesIndef => {}
+            // Type::StringIndef => {}
+            _ => Err(decode::Error::type_mismatch(Type::String)),
         }
-        Ok(Memo(s))
     }
 }
 
-impl TryFrom<String> for Memo<String> {
-    type Error = String;
-    fn try_from(s: String) -> Result<Self, Self::Error> {
-        if s.as_str().as_bytes().len() > MULTISIG_MEMO_DATA_MAX_SIZE {
-            return Err(format!("Memo size over limit {}", s.as_str().as_bytes().len()));
-        }
-        Ok(Memo(s))
+/// A memo contains a human readable portion and/or a machine readable portion.
+/// It is meant to be a note regarding a message, transaction, info or any
+/// type that requires meta information.
+#[derive(Default, Clone, Debug, PartialOrd, Eq, PartialEq)]
+pub struct Memo<const MAX_LENGTH: usize = MULTISIG_MEMO_DATA_MAX_SIZE> {
+    inner: Vec<MemoInner<MAX_LENGTH>>,
+}
+
+impl<const M: usize> Memo<M> {
+    pub fn new() -> Self {
+        Self::default()
+    }
```

Comment:
Would this allow the creation of a `Memo` with an empty `inner`?

We need to enforce that `The memo itself CANNOT be empty and MUST contain at least one entry.`

Reply:
Removed `Default` and `new()`.